src/tools/crossgen2/crossgen2/Program.cs

```diff
@@ -163,182 +159,119 @@ private int Run()
+                        if (singleMethod != null)
                         {
-                            referenceableModules.Add(typeSystemContext.GetModuleFromPath(referenceFile));
+                            // Compiling just a single method
+                            compilationGroup = new SingleMethodCompilationModuleGroup(module, singleMethod);
+                            compilationRoots.Add(new SingleMethodRootProvider(singleMethod));
                         }
-                        catch { } // Ignore non-managed pe files
-                    }
-
-                    ProfileDataManager profileDataManager = new ProfileDataManager(logger, referenceableModules);
-
-                    CompilationModuleGroup compilationGroup;
-                    List<ICompilationRootProvider> compilationRoots = new List<ICompilationRootProvider>();
-                    if (singleMethod != null)
-                    {
-                        // Compiling just a single method
-                        compilationGroup = new SingleMethodCompilationModuleGroup(singleMethod);
-                        compilationRoots.Add(new SingleMethodRootProvider(singleMethod));
-                    }
-                    else
-                    {
-                        // Either single file, or multifile library, or multifile consumption.
-                        EcmaModule entrypointModule = null;
-                        foreach (var inputFile in typeSystemContext.InputFilePaths)
+                        else
                         {
-                            EcmaModule module = typeSystemContext.GetModuleFromPath(inputFile.Value);
-
-                            if (module.PEReader.PEHeaders.IsExe)
-                            {
-                                if (entrypointModule != null)
-                                    throw new Exception("Multiple EXE modules");
-                                entrypointModule = module;
-                            }
+                            // Compiling all the methods in an assembly
+                            compilationGroup = new ReadyToRunSingleAssemblyCompilationModuleGroup(
+                                typeSystemContext, module, referenceableModules, _commandLineOptions.CompileBubbleGenerics,
+                                _commandLineOptions.Partial ? profileDataManager : null);
+                            compilationRoots.Add(new ReadyToRunRootProvider(module, profileDataManager));
                         }
 
-                        List<EcmaModule> inputModules = new List<EcmaModule>();
+                        string inputFileName = referenceableSimpleNameToFileName[module.Assembly.GetName().Name];
+                        string outputFileName = _inputFilePaths.Count > 1 ? Path.Combine(_commandLineOptions.OutputDirectory.FullName, Path.GetFileName(inputFileName)) : _commandLineOptions.OutputFilePath.FullName;
```

Comment:
Hmm, silently switching between using output directory and output file name according to the number of input modules makes me echo some of Michal's concerns, that is indeed somewhat unfortunate. I think that, at the very least, we should:
1) Support "output path and no output file specified" for 1 input path;
2) Check that no explicit output file name is specified in case of multiple input files (at least until we start supporting composite R2R compilation).